src/bgw/scheduler.c

```diff
@@ -325,19 +330,124 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
 	}
 }
 
-static List *
-initialize_scheduled_jobs_list(MemoryContext mctx)
+static void
+terminate_and_cleanup_job(ScheduledBgwJob *sjob)
 {
-	List	   *jobs = bgw_job_get_all(sizeof(ScheduledBgwJob), mctx);
-	ListCell   *lc;
+	if (sjob->handle != NULL)
+	{
+		TerminateBackgroundWorker(sjob->handle);
+		WaitForBackgroundWorkerShutdown(sjob->handle);
+		pfree(sjob->handle);
+	}
+	sjob->may_need_mark_end = false;
+	worker_state_cleanup(sjob);
+}
 
-	foreach(lc, jobs)
+/*
+ *  Update the given job list with whatever is in the bgw_job table. For overlapping jobs,
+ *  copy over any existing scheduler info from the given jobs list.
+ *  Assume that both lists are ordered by job ID.
+ *  Note that this function call will destroy cur_jobs_list and return a new list.
+ */
+List *
+update_scheduled_jobs_list(List *cur_jobs_list, MemoryContext mctx)
+{
+	List	   *new_jobs = bgw_job_get_all(sizeof(ScheduledBgwJob), mctx);
+	ListCell   *new_ptr = list_head(new_jobs);
+	ListCell   *cur_ptr = list_head(cur_jobs_list);
+
+	while (cur_ptr != NULL)
 	{
-		ScheduledBgwJob *sjob = lfirst(lc);
+		ScheduledBgwJob *new_sjob;
+		ScheduledBgwJob *cur_sjob;
+
+		if (new_ptr == NULL)
```

Comment:
move this up into the while conditional?

Reply:
done